File: JerryLi/234_Palindrome Linked List.java
/*
Given a singly linked list, determine if it is a palindrome.

Follow up:
Could you do it in O(n) time and O(1) space?
*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head==null || head.next==null) return true;
        /*
        Two pointers: fast run 2 step, and slow run 1 step;
        when fast run over the list, slow just reach the middle;
        rev pointer follow slow pointer, and reverse the direction of list link.
        */
        ListNode rev=null, fast=head, slow=head;
        boolean result = true;
        while (fast!=null && fast.next!=null){
            fast = fast.next.next;
            if (rev == null){
                rev = slow;
                slow = slow.next;
            } else {
                ListNode tmp = slow;
                slow = slow.next;
                tmp.next = rev;
                rev = tmp;
            }
        }
        if (fast != null) slow = slow.next;
        while (slow != null){
            if (slow.val != rev.val){
                result = false;
                break;
            }
            slow = slow.next;
            rev = rev.next;
        }
        return result;
    }
}

File: JackChen/linked_list/234. Palindrome Linked List.py
"""
Given a singly linked list, determine if it is a palindrome.

Follow up:
Could you do it in O(n) time and O(1) space?
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        rev = None
        fast = head
        while fast and fast.next:
            fast = fast.next.next
            rev, rev.next, head = head, rev, head.next
        head2 = head.next if fast else head
        result = True
        
        while rev:
            result = result and rev.val == head2.val
            head, head.next, rev = rev, head, rev.next
            head2 = head2.next
        return result
